src/qdisc_funcs.sh

```diff
@@ -0,0 +1,636 @@
+            case "$num_tiers $tier" in
+                "1 1") echo "$rate";;
+                "3 1") echo "$prio_rate";;
+                "3 2") echo "$be_rate";;
+                "3 3") echo "$bk_rate";;
+                *)
+                    echo "$rate"
+                    sqm_warn "Unhandled args to flow_rate \"$num_tiers $tier\" in tier_shaper_$SHAPER(). Using defaults.";;
+            esac
+            return;;
+        default_tier)
+            case "$num_tiers" in
+                "1") echo "1";;
+                "3") echo "2";;
+                *)
+                    echo "$(expr $num_tiers - 1)"
+                    sqm_warn "Unhandled args to default_tier \"$num_tiers\" in tier_shaper_$SHAPER(). Using defaults.";;
+            esac
+            return;;
+        tier_args)
+            ;;
+        *)
+            sqm_error "Unknown function \"$func\" in tier_shaper_$SHAPER()."
+            return 1;;
+    esac
+
+    local lq="quantum `get_htb_quantum $IFACE $rate`"
+    local burst="`get_htb_burst $IFACE $rate`"
+    local args="$SHAPER $lq $burst $(get_htb_adsll_string)"
+    local args31="$SHAPER $lq $(get_htb_adsll_string)"
+    local ret=""
+
+    case "$num_tiers $tier" in
+        "1 root") ret="$(get_stab_string) $SHAPER";;
+        "1 0") ret="$args rate ${rate}kbit ceil ${rate}kbit";;
+        "1 1") ret="$args rate ${rate}kbit ceil ${rate}kbit prio 0";;
+    esac
+
+    case "$num_tiers $tier" in
+        "3 root") ret="$(get_stab_string) $SHAPER";;
+        "3 0") ret="$args rate ${rate}kbit ceil ${rate}kbit";;
+        "3 1")
+            case "$dir" in
+                "ingress") ret="$args31 rate 32kbit ceil ${prio_rate}kbit prio 1";;
+                "egress") ret="$args31 rate 128kbit ceil ${prio_rate}kbit prio 1";;
+            esac;;
+        "3 2") ret="$args rate ${be_rate}kbit ceil ${be_ceil}kbit prio 2";;
+        "3 3") ret="$args rate ${bk_rate}kbit ceil ${be_ceil}kbit prio 3";;
+    esac
+
+    [ -n "$ret" ] && echo "$ret" && return
+
+    sqm_error "Unhandled args \"$dir $num_tiers $tier\" to $func in tier_shaper_$SHAPER()."
+    return 1
+}
+
+
+tier_shaper_hfsc() {
+
+    local func=$1
+    local dir=$2
+    local num_tiers=$3
+    local tier=$4
+    local rate
+
+    case $dir in
+        ingress)
+            rate="$DOWNLINK";;
+        egress)
+            rate="$UPLINK";;
+        *)
+            sqm_error "Unknown direction \"$dir\" in tier_shaper_$SHAPER()."
+            return 1;;
+    esac
+
+# Parameter comments below from Eric Luehrsen:
+
+# Link share (virtual time) tuning is only rough as borrowing will occur.
+# However in saturation, each class is guaranteed (real time) minimum of 15%.
+# This allows HFSC to do its work but avoids some of its vitrual time quirks.
+
+    local rate_c=$(( ${rate} * 50 / 49 ))
+    local rate_r=$(( ${rate} * 1 / 7 ))
+    local rate_1=$(( ${rate} * 2 / 7 ))
+    local rate_2=$(( ${rate} * 2 / 7 ))
+    local rate_3=$(( ${rate} * 3 / 7 ))
+
```

Comment:
I did not work on nxt_routed_hfsc. It was unique in being for IP4 NAT. It would attempt to get the 'real' IP addresses for fq_codel to chew on.